### ISSUE TITLE
Problem when iterating over HashGetAll() and then calling HashDelete()

### ISSUE THREAD

**Issue by styfle** [StackExchange/StackExchange.Redis#862]
I admit that I could be doing something wrong so forgive me if this is not a bug in the lib.

I have a single Redis server with the following configuration:

```cs
var options = new ConfigurationOptions
{
    EndPoints = {
        { ConfigurationManager.AppSettings["RedisHost"], int.Parse(ConfigurationManager.AppSettings["RedisPort"]) }
    },
    Password = ConfigurationManager.AppSettings["RedisPassword"]
};
var connection = ConnectionMultiplexer.Connect(options);
var db = connection.GetDatabase(1);
```

I have method `ClearRegion` that is attempting to delete the entire hash (note that `Region` is an enum).

The following is **not** deleting keys:

```cs
public void ClearRegion(Region region)
{
    db.HashGetAll(region.ToString()).Select(entry =>
        db.HashDelete(region.ToString(), entry.Name)
    );
}
```

After some refactoring, the following is deleting keys as expected:

```cs
public void ClearRegion(Region region)
{
    var entries = db.HashGetAll(region.ToString());
    foreach (var entry in entries)
    {
        var success = db.HashDelete(region.ToString(), entry.Name);
    }
}
```

So is this a bug or am I doing it wrong? 😄

**Comment by mgravell** [comment]
> So is this a bug or am I doing it wrong?

The latter.

It is a common misconception, but:

    db.HashGetAll(region.ToString()).Select(entry =>
        db.HashDelete(region.ToString(), entry.Name)
    );

**does not** actually iterate the data; it *returns* something that **when that something is iterated** : iterates the inner data. So basically, you didn't call `HashDelete` ever. You can see this by doing instead:

    db.HashGetAll(region.ToString()).Select(entry =>
        Console.WriteLine(entry);
    );

Basically: use a `foreach` loop here.

**Comment by styfle** [comment]
Oh you're right, I should have known that!

Classic mistake on my part, my apologies 😅 
(I did have 2 other people review the code before I submitted this bug so shame on them too haha)

**Comment by mgravell** [comment]
no problem :) we've all been there
